### PR TITLE
feature/get all users from keycloack

### LINKED ISSUE #267: Get users from keycloack
Implement getall method to get all users from keycloack database.

### PR DESCRIPTION
**Pull request by Mohammad96Assaf**
close #267

### CODE REVIEW

File: internal/server/http/handler/v1/user/handler.go [hunk 1]
```diff
@@ -187,9 +188,41 @@ func parseURL(rawURL string) (*url.URL, error) {
 // @Param			limit	query		string	false	"Limit"
 // @Router			/v1/user [get]
 // @Security		Keycloak
-func GetAllUsers(_ service.AuthService) fiber.Handler {
+func GetAllUsers(svc service.AuthService) fiber.Handler {
 	return func(c *fiber.Ctx) error {
-		return c.SendStatus(fiber.StatusNotImplemented)
+		page := c.Query("page")
```
Comment (doriengr) [inline internal/server/http/handler/v1/user/handler.go:193]: Are you currently using page and limit somewhere? We need those parameters if we want to use pagination, what we are not doing (yet).

File: internal/service/domain/auth/user.go [hunk 2]
```diff
@@ -66,3 +66,12 @@ func (s *AuthService) LogoutRequest(ctx context.Context, logoutRequest *domain.L
 
 	return nil
 }
+
+func (s *AuthService) GetAllUsers(ctx context.Context) ([]*domain.User, error) {
```
Comment (doriengr) [inline internal/service/domain/auth/user.go:70]: In the other services we call this method just »GetAll«. 
To make it uniform everywhere, I would suggest that we name it GetAll here as well? Would that be okay?

File: internal/server/http/handler/v1/user/handler.go [hunk 3]
```diff
@@ -187,9 +188,41 @@ func parseURL(rawURL string) (*url.URL, error) {
...
+		ctx := c.Context()
+
+		users, err := svc.GetAllUsers(ctx)
+		if err != nil {
+			return errorhandler.HandleError(service.NewError(service.InternalError, errors.Wrap(err, "failed to get all users").Error()))
+		}
+		response := make([]entities.UserResponse, len(users))
+
+		for i, user := range users {
+			response[i] = entities.UserResponse{
```
Comment (doriengr) [inline internal/server/http/handler/v1/user/handler.go:212]: Maybe we should use the http mapper to map the data?

File: internal/service/domain/auth/user_test.go [hunk 4]
```diff
@@ -291,3 +291,61 @@ func TestLogoutRequest(t *testing.T) {
 		assert.Error(t, err)
 	})
 }
+
+func TestGetAllUsers(t *testing.T) {
+	t.Run("should return all users successfully", func(t *testing.T) {
+		// given
```
Comment (doriengr) [inline internal/service/domain/auth/user_test.go:297]: Maybe add a test case »should return empty slice when user are found«? Same goes for the handler.

File: internal/storage/auth/keycloak/user.go [hunk 5]
```diff
@@ -83,21 +83,50 @@ func (r *UserRepository) RemoveSession(ctx context.Context, refreshToken string)
...
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get users from Keycloak")
+	}
+
+	var allUsers []*entities.User
+
+	for _, kcUser := range users {
+		user, err := keyCloakUserToUser(kcUser)
+		if err != nil {
+			continue
```
Comment (doriengr) [inline internal/storage/auth/keycloak/user.go:103]: Why should it be continued if an error is happening?

File: internal/storage/auth/keycloak/user_test.go [hunk 6]
```diff
@@ -297,3 +297,68 @@ func TestKeyCloakUserRepo_UserToKeyCloakUser(t *testing.T) {
...
+		suite.EnsureUserExists(t, user2)
+
+		// when
+		users, err := userRepo.GetAll(context.Background())
+
+		// then
+		assert.NoError(t, err)
+		assert.NotNil(t, users)
+		assert.GreaterOrEqual(t, len(users), 2)
+		assert.True(t, containsUser(users, *user1), "user1 should be in the list")
```
Comment (doriengr) [inline internal/storage/auth/keycloak/user_test.go:333]: I am not sure, but can you maybe use `assert.Contain`?

**Review by doriengr — CHANGES_REQUESTED** [review]

File: internal/server/http/handler/v1/user/handler.go (hunk 1 above)
Comment (Mohammad96Assaf) [reply internal/server/http/handler/v1/user/handler.go:193]: Actually, I wanted to ask the group if we need it in the handler. But I will delete it for now since we don't use it 👍

File: internal/service/domain/auth/user_test.go (hunk 4 above)
Comment (Mohammad96Assaf) [reply internal/service/domain/auth/user_test.go:297]: I think, when **no** users are found.

File: internal/storage/auth/keycloak/user.go (hunk 5 above)
Comment (Mohammad96Assaf) [reply internal/storage/auth/keycloak/user.go:103]: Continuing would be similar to skipping users that are nil, we don't need to add them to the response list.
It's the same as saying: `if user != nil { allUsers = append(allUsers, user) }`

But what do you suggest, should we break the loop if the conversion from a Keycloak user to user fails and return an error ?

File: internal/storage/auth/keycloak/user_test.go (hunk 6 above)
Comment (Mohammad96Assaf) [reply internal/storage/auth/keycloak/user_test.go:333]: That's the reason why I implemented the containsUser function. The assert.Contains method checks if the pointer (memory reference) is included or not, which doesn't work in my case

File: internal/storage/auth/keycloak/user.go (hunk 5 above)
Comment (doriengr) [reply internal/storage/auth/keycloak/user.go:103]: I would say so, yes ...

**Comment by doriengr** [comment]
Since I believe that the frontend credentials must be used, the error still needs to be fixed here. I hope that is correct, that the frontend credentials must always be used in Swagger, right? @choffmann 
image

**Review by doriengr — APPROVED** [review]
